### PR TITLE
indexer: quick object snapshot and fix object history

### PR DESCRIPTION
**Pull request by gegaowp**
## Description 

main changes are:
- instead of updating object snapshot at epoch boundary, now we update object snapshot following a sliding window of checkpoints, so that each update will not take too long
- fix object history, prev it skipped intermediate object versions, which is correct for `objects` which only tracks latest versions, but not the case for `objects_history`, which is supposed to track all versions
- move advance epoch before checkpoint persisting, so that checkpoint can always be the watermark, in case of pod crashing and resuming, the advance epoch process will be resumed properly too

## Test Plan 

local run and verify that
- objects_history table has all intermediate versions & checkpoints
- object snapshot can be update properly per the objects_history

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-indexer/src/indexer_reader.rs [hunk 1]
```diff
@@ -1637,6 +1637,31 @@ impl IndexerReader {
                 )))?;
         Ok(TreasuryCap::try_from(treasury_cap_obj_object)?.total_supply)
     }
+
+    pub fn get_consistent_read_range(&self) -> Result<(i64, i64), IndexerError> {
```
Comment (gegaowp) [inline crates/sui-indexer/src/indexer_reader.rs:1641]: @oxade this is the available range function

File: crates/sui-indexer/src/indexer_reader.rs [hunk 2]
```diff
@@ -1637,6 +1637,31 @@ impl IndexerReader {
...
+        let latest_checkpoint_sequence = self
+            .run_query(|conn| {
+                checkpoints::table
+                    .select(checkpoints::sequence_number)
+                    .order(checkpoints::sequence_number.desc())
+                    .first::<i64>(conn)
```
Comment (emmazzz) [inline crates/sui-indexer/src/indexer_reader.rs:1647]: n00b q: can we use `max` here? i.e. `select(max(checkpoints::sequence_number)).first::<i64>(conn)`. Same for the query below.

File: crates/sui-indexer/src/store/pg_indexer_store_v2.rs [hunk 3]
```diff
@@ -74,6 +74,35 @@ const PG_COMMIT_PARALLEL_CHUNK_SIZE_PER_DB_TX: usize = 500;
...
+const UPDATE_OBJECTS_SNAPSHOT_QUERY: &str = r"
+INSERT INTO objects_snapshot (object_id, object_version, object_status, object_digest, checkpoint_sequence_number, owner_type, owner_id, object_type, serialized_object, coin_type, coin_balance, df_kind, df_name, df_object_type, df_object_id)
+SELECT object_id, object_version, object_status, object_digest, checkpoint_sequence_number, owner_type, owner_id, object_type, serialized_object, coin_type, coin_balance, df_kind, df_name, df_object_type, df_object_id
+FROM (
+    SELECT *,
+           ROW_NUMBER() OVER (PARTITION BY object_id ORDER BY object_version DESC) as rn
+    FROM objects_history
+    WHERE checkpoint_sequence_number >= $1 AND checkpoint_sequence_number < $2
+) as subquery
+WHERE rn = 1
```
Comment (emmazzz) [inline crates/sui-indexer/src/store/pg_indexer_store_v2.rs:89]: Is this a fancy way of doing "get all objects with their greatest versions"? Having a comment explaining here would be great. Thanks!

**Comment by XoL1507** [comment]
should i change to read and write?

File: crates/sui-indexer/src/store/pg_indexer_store_v2.rs (hunk 3 above)
Comment (gegaowp) [reply crates/sui-indexer/src/store/pg_indexer_store_v2.rs:89]: yes, `rn` is a generated rank from ROW_NUMBER(), which is ordered rank based on desc object version.

Comment (gegaowp) [reply crates/sui-indexer/src/store/pg_indexer_store_v2.rs:91]: added the comments.

File: crates/sui-indexer/src/store/pg_indexer_store_v2.rs [hunk 4]
```diff
@@ -865,6 +950,38 @@ impl IndexerStoreV2 for PgIndexerStoreV2 {
...
 
+    async fn persist_object_snapshot(&self) -> Result<(), IndexerError> {
+        let latest_cp = self
+            .get_latest_tx_checkpoint_sequence_number()?
+            .unwrap_or_default();
+        let latest_snapshot_cp = self
+            .get_latest_object_snapshot_checkpoint_sequence_number()?
+            .unwrap_or_default();
+
+        if latest_cp <= latest_snapshot_cp + self.object_snapshot_max_checkpoint_lag as u64 {
```
Comment (emmazzz) [inline crates/sui-indexer/src/store/pg_indexer_store_v2.rs:961]: Can you add some comments explaining the relationship between `latest_cp`, `latest_snapshot_cp`, `object_snapshot_max_checkpoint_lag`. It's slightly hard for someone who doesn't know this code well to understand why we are returning here, and the logic in general. Also `object_snapshot_min_checkpoint_lag` seems to be only used for determining the window? What if we store the window length directly?

Comment (gegaowp) [reply crates/sui-indexer/src/store/pg_indexer_store_v2.rs:970]: > What if we store the window length directly?
yes, we can do constants of either max_lag & min_lag, or max_lag & window_size, I think min_lag is more meaningful for the clients, b/c this is the min size of the consistent read range, while window size is more of impl details

**Review by emmazzz — APPROVED** [review]

**Pull request merged**